filemanager.go

```diff
@@ -0,0 +1,358 @@
+
+	if err := m.snapshotRemote(ctx, dest); err != nil {
+		return err
+	}
+
+	if err := m.snapshotLocal(source); err != nil {
+		return err
+	}
+
+	updates, err := m.filterUploads()
+
+	if err != nil {
+		return err
+	}
+
+	for _, uploadFile := range updates.files {
+		if uploadFile.IsDir {
+			continue
+		}
+
+		if err := m.upload(ctx, uploadFile, source, dest); err != nil {
+			return err
+		}
+	}
+
+	if m.del {
+		deletes := m.filterDeletes()
+		for _, remotePath := range deletes {
+			if err := m.deleteRemote(remotePath); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// Purge deletes all files in the given path.
+func (m *Manager) Purge(ctx context.Context, remotePath string) error {
+
+	if err := m.snapshotRemote(ctx, remotePath); err != nil {
+		return err
+	}
+
+	deletes := m.filterDeletes()
+	for _, deletePath := range deletes {
+		if err := m.deleteRemote(deletePath); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// List prints the files in the filemanager.
+func (m *Manager) List(ctx context.Context, remotePath string) error {
+
+	if err := m.snapshotRemote(ctx, remotePath); err != nil {
+		return err
+	}
+
+	for _, remoteFile := range m.remoteFiles.files {
+		fmt.Printf("%s\n", remoteFile.Path)
+	}
+	return nil
+}
+
+// filterDeletes returns the files that need to be deleted.
+func (m *Manager) filterDeletes() []string {
+
+	deletes := []string{}
+
+	for remoteRelativeName, remoteFile := range m.remoteFiles.files {
+		if _, ok := m.localFiles.get(remoteRelativeName); !ok {
+			deletes = append(deletes, remoteFile.Path)
+		}
+	}
+	sort.Sort(sort.Reverse(sort.StringSlice(deletes)))
+	return deletes
+}
+
+// filterUploads returns the files that need to be uploaded.
+func (m *Manager) filterUploads() (*fileMap, error) {
+
+	uploads := fileMap{
+		files: make(map[string]*fileInfo),
+	}
+
+	for localRelativeName, localFile := range m.localFiles.files {
+		if localFile.IsDir {
+			continue
+		}
+
+		remoteFile, ok := m.remoteFiles.get(localRelativeName)
+		if !ok {
+			uploads.add(localRelativeName, localFile)
+			continue
+		}
+
+		if localFile.Size != remoteFile.Size {
+			uploads.add(localRelativeName, localFile)
+			continue
+		}
+
+		if localDigest, err := getFileDigest(localFile.Path); err != nil {
+			return nil, err
+		} else if localDigest != remoteFile.Digest {
+			uploads.add(localRelativeName, localFile)
+		}
+	}
+	return &uploads, nil
+}
+
+// snapshotLocal returns a channel which receives the infos of the files under the given basePath.
+func (m *Manager) snapshotLocal(path string) error {
+
+	m.localFiles.files = make(map[string]*fileInfo)
+
+	basePath := filepath.ToSlash(filepath.Clean(path))
+	stat, err := os.Stat(basePath)
+
+	if err != nil {
+		return err
+	}
+
+	if !stat.IsDir() {
+		return fmt.Errorf("the path %s is not a directory", basePath)
+	}
+
+	err = filepath.Walk(basePath, func(p string, stat os.FileInfo, err error) error {
```

Comment:
If you rename `path` in definition of `snapshotLocal` to `basePath` or `rootPath`, then you can use `path` instead of `p`. That would make it a bit easier to read.